File: Clojure/treap.clj
(ns treap
  (:refer-clojure
    :exclude [print remove rand]))

(defn- rand []
  (rand-int Integer/MAX_VALUE))

(defrecord node [x y s l r])

(defn- new-node [x]
  (->node x (rand) 1 nil nil))

(defn size [{s :s}]
  (or s 0))

(defn join
  ([{y1 :y s1 :s r1 :r :as lf}
    {y2 :y l2 :l s2 :s :as rg}]
    (cond
      (nil? lf) rg
      (nil? rg) lf
      (< y1 y2)
        (assoc lf
          :r (join r1 rg)
          :s (inc s1))
      :else
        (assoc rg
          :l (join lf l2)
          :s (inc s2))))
  ([a b & c]
    (reduce join (join a b) c)))

(defn split [{:keys [x l r] :as p} v]
  (cond
    (nil? p)
      [nil nil]
    (< v x)
      (let [[lf rg] (split l v)]
        [lf
         (assoc p
           :l rg
           :s (+ (size rg) 1 (size r)))])
    :else
      (let [[lf rg] (split r v)]
        [(assoc p
           :r lf
           :s (+ (size l) 1 (size lf)))
         rg])))

(defn insert [p v]
  (let [[l r] (split p v)]
    (join l (new-node v) r)))

(defn remove [{:keys [x s l r] :as p} v]
  (cond
    (nil? p) nil
    (< v x)
      (let [l (remove l v)]
        (assoc p
          :l l
          :s (+ (size l) 1 (size r))))
    (> v x)
      (let [r (remove r v)]
        (assoc p
          :r r
      	 	  :s (+ (size l) 1 (size r))))
    :else (join l r)))

(defn remove-range [p l r]
  (let [[a b] (split p r)]
    (let [[x y] (split a (- l 1))]
      (join x b))))

(defn print [{:keys [x l r] :as p}]
  (lazy-seq
    (if (nil? p)
      nil
      (concat (print l)
              (list x)
              (print r)))))

(defn make-treap [a]
  (loop [t nil [f & r] a]
    (if (nil? f)
      t
      (recur (insert t f) r))))

File: Go/treap.go
package main

import (
	"cmp"
	"math/rand/v2"
)

type Node[T cmp.Ordered] struct {
	Value       T
	heapValue   int
	Left, Right *Node[T]
}

func NewNode[T cmp.Ordered](x T) *Node[T] {
	return &Node[T]{
		Value:     x,
		heapValue: rand.Int(),
	}
}

func (left *Node[T]) join(right *Node[T]) *Node[T] {
	if left == nil {
		return right
	} else if right == nil {
		return left
	} else if left.heapValue < right.heapValue {
		left.Right = left.Right.join(right)
		return left
	} else {
		right.Left = left.join(right.Left)
		return right
	}
}

func (node *Node[T]) split(value T) (*Node[T], *Node[T]) {
	if node == nil {
		return nil, nil
	} else if cmp.Compare(value, node.Value) < 0 {
		leftLeft, leftRight := node.Left.split(value)
		node.Left = leftRight
		return leftLeft, node
	} else {
		rightLeft, rightRight := node.Right.split(value)
		node.Right = rightLeft
		return node, rightRight
	}
}

func (node *Node[T]) insert(value T) *Node[T] {
	left, right := node.split(value)
	return left.join(NewNode(value)).join(right)
}

func (node *Node[T]) print() {
	if node == nil {
		return
	}
	node.Left.print()
	println(node.Value)
	node.Right.print()
}

func main() {

	var t *Node[int]

	for i := range 100_000 {
		t = t.insert(i)
	}

	t.print()
}
